### PR TITLE
Add type parameter to SimSpec to avoid free variables escaping scope

### LINKED ISSUE #304: Clean up bound/free variable scoping
Currently the verifier is not carefully handling the free variables introduced to represent the initial program state for symbolic execution. In particular the body of the 'SimSpec' underlying the 'AbstractDomain' is accessed directly rather than being bound via 'bindSpec'. This creates two problems: domains may include free variables which are unbound once interpreted in a subsequent slice, and domains may include free variables which are bound, but should actually be free (i.e. in a loop each iteration requires different formal variables to represent the initial state).

### PR DESCRIPTION
**Pull request by danmatichuk**
Types that are intended to contain bound variables that are to be instantiated later (i.e. via bindSpec) now have an additional type parameter that is used to track the scope of these variables.

This allows us to more easily track where values might escape their scope and ensure that they have been properly been transformed and phrased over an appropriate set of bound variables.

This change makes it clear where the current scope violation is happening, (see `Pate.Verification.Widening.abstractOverVars`) where we "unsafely" coerce a domain from one scope to another. This is a nontrivial problem in general is exactly where we need to add careful treatment for the stack pointer.

fixes #304

### CODE REVIEW

**Review by travitch — APPROVED** [review]

File: src/Pate/Verification/Widening.hs [hunk 1]
```diff
@@ -125,37 +126,41 @@ makeFreshAbstractDomain _bak _bundle preDom (ReturnNode fPair) = do
...
+                 return $ recordMiscAnalysisError (freshDomain gr to postSpec') to (Text.pack msg')
+            Widen _ _ d'' -> do
+              postSpec' <- abstractOverVars bundle from to postSpec d''
+              return (freshDomain gr to postSpec')
 
     -- have visited this location at least once before
-    Just d' ->
-     do md <- widenPostcondition bak bundle d d'
+    Just postSpec -> do
+      (asm, d') <- liftIO $ PS.bindSpec sym (PS.bundleOutVars bundle) postSpec
```
Comment (travitch) [inline src/Pate/Verification/Widening.hs:161]: Lets comment what these `bindSpec` calls are doing (and, more importantly, why)

File: src/Pate/Monad.hs [hunk 2]
```diff
@@ -363,28 +366,30 @@ lookupArgumentNames pp = do
...
   binCtx <- getBinCtx @bin
   let baseMem = MBL.memoryImage $ PMC.binary binCtx
   mem <- withSymIO $ \sym -> MT.initMemTrace sym baseMem (MM.addrWidthRepr (Proxy @(MM.ArchAddrWidth arch)))
   argNames <- lookupArgumentNames blocks
   regs <- MM.mkRegStateM (\r -> unconstrainedRegister argNames r)
-  return $ SimVars regs (SimState mem (MM.mapRegsWith (\_ -> PSR.macawVarEntry) regs))
+  return $ SimBoundVars regs (SimState mem (MM.mapRegsWith (\_ -> PSR.macawVarEntry) regs))
 
 
 withFreshVars ::
```
[the remark this answers is not in the document]
Comment (danmatichuk) [reply src/Pate/Monad.hs:384]: I added some comments above the 'VarScope' kind which tries to capture what the variable scope parameter is for.

File: src/Pate/SimState.hs [hunk 3]
```diff
@@ -271,90 +281,123 @@ rebindWithFrame' ::
...
-  , specBody :: f
-  }
-
 
-specVarsO :: SimSpec sym arch f -> SimVars sym arch PBi.Original
-specVarsO spec = PPa.pOriginal $ specVars spec
+-- | Proof that the interpretation of 'f' is independent of the scope variable.
+-- This is necessary since we do want to convert between scopes as a result of binding.
+class Scoped f where
+  unsafeCoerceScope :: forall v v'. f v -> f v'
```
[the remark this answers is not in the document]
Comment (danmatichuk) [reply src/Pate/SimState.hs:288]: I added some more documentation and a TODO for eventually adding a "safe" variant that checks the intended scope invariant.

File: src/Pate/SimState.hs [hunk 4]
```diff
@@ -271,90 +281,123 @@ rebindWithFrame' ::
...
 
-specVarsP :: SimSpec sym arch f -> SimVars sym arch PBi.Patched
-specVarsP spec = PPa.pPatched $ specVars spec
+data SimSpec sym arch (f :: VarScope -> DK.Type) where
+  SimSpec :: forall sym arch f v.
```
[the remark this answers is not in the document]
Comment (danmatichuk) [reply src/Pate/SimState.hs:291]: fixed

File: src/Pate/Verification/Widening.hs [hunk 5]
```diff
@@ -256,17 +271,17 @@ withPredomain bundle preD f = withSym $ \sym -> do
  where
    doPanic = panic Solver "withPredomain" ["Online solving not enabled"]
 
-widenPostcondition :: forall sym arch solver t st fs bak.
+widenPostcondition :: forall sym arch solver t st fs bak v v'.
```
[the remark this answers is not in the document]
Comment (danmatichuk) [reply src/Pate/Verification/Widening.hs:274]: Nope, this was an accident.

File: src/Pate/Verification/Widening.hs (hunk 1 above)
Comment (danmatichuk) [reply src/Pate/Verification/Widening.hs:161]: Added some comments

**Pull request merged**